### PR TITLE
Fix an issue with zero entry in split TimeSeriesProperty

### LINKED ISSUE #21836: Some sample log has zero entry after FilterEvents
Here is the script

```
LoadEventNexus(Filename='/SNS/REF_M/shared/ADARA.Test.Data.2018/REF_M_28131.nxs.h5',
               OutputWorkspace='test31')
               
GenerateEventsFilter(InputWorkspace='test31', OutputWorkspace='filter', InformationWorkspace='filter_info', LogName='SF1',
                     MinimumLogValue=0.9, MaximumLogValue=1.1, LogBoundary='Left')
FilterEvents(InputWorkspace='test31', SplitterWorkspace='filter', OutputWorkspaceBaseName='testtest',
             OutputWorkspaceNames='testtest_unfiltered,testtest_0')

# Compare the CurrentSP log to the one from this workspace
FilterByLogValue(InputWorkspace='test31', OutputWorkspace='old', LogName='SF1',
                 MinimumValue=0.9, MaximumValue=1.1, LogBoundary='Left')

```

In addition, input workspaces with logs with zero entries should be treated properly.
Currently, a time series with zero entries in the input workspace will cause Mantid to seg fault.

### Expected behavior

Log 'SF1' shall have at least 1 value (the last one) in output workspace from FilterEvents.

### Actual behavior

In workspace 'testtest', log 'SF1' has no entry.

### PR DESCRIPTION
**Pull request by wdzhou**
Method splitByTimeVector may generate split TimeSeriesProperty with zero entry.  The behaviour shall be changed such that there shall be at least one entry in each split TimeSeriesProperty instance.

**Report to:** [mdoucet ]. 

**To test:**

```
LoadEventNexus(Filename='/SNS/REF_M/shared/ADARA.Test.Data.2018/REF_M_28131.nxs.h5',
               OutputWorkspace='test31')
               
GenerateEventsFilter(InputWorkspace='test31', OutputWorkspace='filter', InformationWorkspace='filter_info', LogName='SF1',
                     MinimumLogValue=0.9, MaximumLogValue=1.1, LogBoundary='Left')
FilterEvents(InputWorkspace='test31', SplitterWorkspace='filter', OutputWorkspaceBaseName='testtest',
             OutputWorkspaceNames='testtest_unfiltered,testtest_0')
```
In output workspace testtest_0, sample log CurrentSP shall have 1 entry.

Fixes #21836 . 

Does this update require release notes?
- [X] Yes
- [ ] No

Release note is added to docs/source/release/framework.rst

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)?
- [ ] Are the unit tests small and test the class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] Do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Review by mdoucet — APPROVED** [review]
I tested it with a data set that previously failed and it worked as expected.

**Comment by quantumsteve** [comment]
Please remove old, commented out code.

File: Framework/Algorithms/inc/MantidAlgorithms/FilterEvents.h [hunk 1]
```diff
@@ -164,10 +164,11 @@ class DLLExport FilterEvents : public API::Algorithm {
...
 
-  void splitDoubleTimeSeriesLogs(
-      const std::vector<Kernel::TimeSeriesProperty<double> *> &dbl_tsp_vector,
-      std::vector<Types::Core::DateAndTime> &split_datetime_vec,
-      const int max_target_index);
+  //  void splitDoubleTimeSeriesLogs(
+  //      const std::vector<Kernel::TimeSeriesProperty<double> *>
+  //      &dbl_tsp_vector,
+  //      std::vector<Types::Core::DateAndTime> &split_datetime_vec,
+  //      const int max_target_index);
```
Comment (SimonHeybrock) [inline Framework/Algorithms/inc/MantidAlgorithms/FilterEvents.h]: Leftover commented code?

File: Framework/Algorithms/test/FilterEventsTest.h [hunk 2]
```diff
@@ -1379,15 +1382,15 @@ class FilterEventsTest : public CxxTest::TestSuite {
     Kernel::SplittingInterval interval0(t0, t1, 0);
     splitterws->addSplitter(interval0);
 
-    std::cout << "Add splitters: " << t0 << ", " << t1 << ", " << 0 << "\n";
+    // std::cout << "Add splitters: " << t0 << ", " << t1 << ", " << 0 << "\n";
```
Comment (SimonHeybrock) [inline Framework/Algorithms/test/FilterEventsTest.h]: More commented code.

File: Framework/Algorithms/test/FilterEventsTest.h [hunk 3]
```diff
@@ -1379,15 +1382,15 @@ class FilterEventsTest : public CxxTest::TestSuite {
...
 
     // 2. Splitter 1: 3+ ~ 9+ (second pulse)
     t0 = t1;
     t1 = runstart_i64 + pulsedt + tofdt * 9 + tofdt / 2;
     Kernel::SplittingInterval interval1(t0, t1, 1);
     splitterws->addSplitter(interval1);
 
-    std::cout << "Add splitters: " << t0 << ", " << t1 << ", " << 1 << "\n";
+    // std::cout << "Add splitters: " << t0 << ", " << t1 << ", " << 1 << "\n";
```
Comment (SimonHeybrock) [inline Framework/Algorithms/test/FilterEventsTest.h]: .

File: Framework/Kernel/src/TimeSeriesProperty.cpp [hunk 4]
```diff
@@ -566,6 +572,7 @@ void TimeSeriesProperty<TYPE>::splitByTimeVector(
   std::vector<DateAndTime>::iterator splitter_iter;
   splitter_iter = std::lower_bound(splitter_time_vec.begin(),
                                    splitter_time_vec.end(), tsp_time);
+  // tps_time is start time of time series property
```
Comment (SimonHeybrock) [inline Framework/Kernel/src/TimeSeriesProperty.cpp]: `tsp_time`

Also, it does not show up until 20 lines further down. Move the comment there?

File: Framework/Kernel/src/TimeSeriesProperty.cpp [hunk 5]
```diff
@@ -594,17 +599,35 @@ void TimeSeriesProperty<TYPE>::splitByTimeVector(
...
     }
+  } else {
+    // no entry in range is true, which corresponding to the previous case
+    // "already search to the last splitter which is still earlier than first
+    // TSP
+    // entry"
+    ;
+  }
+
+  //
```
Comment (SimonHeybrock) [inline Framework/Kernel/src/TimeSeriesProperty.cpp]: Empty comment line.

File: Framework/Kernel/src/TimeSeriesProperty.cpp [hunk 6]
```diff
@@ -664,8 +688,44 @@ void TimeSeriesProperty<TYPE>::splitByTimeVector(
       split_start_time = split_stop_time;
       split_stop_time = splitter_time_vec[index_splitter + 1];
     }
+
+    ++outer_while_counter;
   } // END-OF-WHILE
 
+  // It was in 'continue search'-while-loop.  But the TSP runs over before
+  // splitters.
```
Comment (SimonHeybrock) [inline Framework/Kernel/src/TimeSeriesProperty.cpp]: What is "It"? I do not understand the comment.

File: Framework/Kernel/test/TimeSeriesPropertyTest.h [hunk 7]
```diff
@@ -796,7 +796,7 @@ class TimeSeriesPropertyTest : public CxxTest::TestSuite {
   /** last splitter is before first entry
    * @brief test_splitByTimeVectorEarlySplitter
    */
-  void test_splitByTimeVectorEarlySplitter() {
+  void Throw_test_splitByTimeVectorEarlySplitter() {
```
Comment (SimonHeybrock) [inline Framework/Kernel/test/TimeSeriesPropertyTest.h]: Is this test disabled? Why?

**Review by SimonHeybrock — CHANGES_REQUESTED** [review]
A couple of small comments:

**Review by SimonHeybrock — APPROVED** [review]

**Pull request merged**